kdumpctl

```diff
@@ -1659,47 +1662,156 @@ _should_reset_crashkernel() {
 	[[ $(kdump_get_conf_val auto_reset_crashkernel) != no ]] && systemctl is-enabled kdump &> /dev/null
 }
 
+set_kdump_test_id()
+{
+	local _id=$1
+	local _kmsg="/dev/kmsg"
+
+	if [[ $DEFAULT_DUMP_MODE == "fadump" ]]; then
+		[[ -c "$_kmsg" ]] && echo "$_id" > "$_kmsg"
+	else
+		KDUMP_COMMANDLINE_APPEND+=" $_id "
+		reload >& /dev/null
+	fi
+
+	if [[ "$?" -ne 0 ]]; then
+		derror "Set kdump test id fail."
+		exit 1
+	fi
+}
+
+# $1: success/fail/pending/manual/clear
+# $2: test id
+set_vmcore_creation_status()
+{
+	local _status=$1
+	local _test_id=$2
+	_dir=$(dirname "$VMCORE_CREATION_STATUS")
+
+	[[ -d "$_dir" ]] || mkdir -p "$_dir"
+	[[ -w "$_dir" ]] || chmod +w "$_dir"
+
+	case "$_status" in
+		pending)
+			echo  "$_status $_test_id" > "$VMCORE_CREATION_STATUS"
+			;;
+		success | fail | manual)
+			sed -E -i "s/^\w+/$_status/" "$VMCORE_CREATION_STATUS"
+			;;
+		clear)
+			rm -f "$VMCORE_CREATION_STATUS"
+			;;
+		*)
+			return
+	esac
+	sync -f "$_dir"
+}
+
+fetch_test_status()
+{
+	local _test_id="$1"
+	local _script="$KDUMP_TMPDIR/fetch.sh"
+
+	cat << EOF > "$_script"
+#!/bin/bash
+dmesg_files=\$(find "\$1" -name "kexec-dmesg.log" -or -name "vmcore-dmesg.txt" 2> /dev/null)
+[[ -z "\$dmesg_files" ]] && exit 2
+dmesg_files=\$(ls -t \$dmesg_files | head -50)
+id_file=\$(grep -l "$_test_id" \$dmesg_files | uniq)
+case \$(echo -n "\$id_file" | wc -w) in
+	0)
+		# Not found files containing the test id
+		exit 2
+		;;
+	*)
+		# Found one or more files containing the test id
+		dir=\$(dirname \$id_file | uniq)
+		case \$(echo -n "\$dir" | wc -w) in
+			1)
+				# The test id files within one folder
+				;;
+			*)
+				# The test id files shouldn't within
+				# multiple folders
+				exit 2
+				;;
+		esac
+		;;
+esac
+if [[ -f "\$dir/vmcore" || -f "\$dir/vmcore.flat" ]]; then
+	# Success
+	exit 0
+else
+	# Fail
+	exit 1
+fi
+EOF
```

Comment:
I'm not really comfortable with creating and running the script as is. For one it creates a dependency for the host, when using a ssh target. For the other writing the script to a temporary file and then running it opens a race window in which attackers could replace the script with their own. Leading to a potential privilege escalation when we don't pay enough attention.

Could we simplify this by storing the status in a file known to both kdumpctl and the initrd, e.g. something like `$path/.$test_id.status`. That would require to run some code in the initrd to parse the `$test_id` from the command line and write the final status to the file. But that would allow to make the code overall much simpler IMHO and still prevents the need to always mount the rootfs in the initrd.

Reply:
Hi Philipp,

> I'm not really comfortable with creating and running the script as is. For one it creates a dependency for the host, when using a ssh target. For the other writing the script to a temporary file and then running it opens a race 

I'm not very clear for the dependency here, do you suggest we don't ask remote server to execute script at all?

> window in which attackers could replace the script with their own. Leading to a potential privilege escalation when we don't pay enough attention.

Agreed, we can get rid of the temporary script file, how about a use of typeset like this: https://stackoverflow.com/questions/22107610/shell-script-run-function-from-script-over-ssh?

> 
> Could we simplify this by storing the status in a file known to both kdumpctl and the initrd, e.g. something like `$path/.$test_id.status`. That would require to run some code in the initrd to parse the `$test_id` from the command line and write the final status to the file. But that would allow to make the code overall much simpler IMHO and still prevents the need to always mount the rootfs in the initrd.

If we want initrd, aka 2nd kernel to access the status file, assuming the status file is independent from vmcore it  is trying to save, this will involve extra kernel module to deal with the fs where the status file is located, and this is what we are trying to avoid, because extra kernel module will increase the kernel failure, preventing vmcore saving, and it is unacceptable.

If the status file is saved adjoining vmcore, then no extra module is needed. Since we still need to access the remote server to check for the status when we back into 1st kernel, so there is no point for creating the status file next to vmcore remotely right?

Please correct me if I misunderstood your approach...

Reply:
I removed the temporary script generation in v2.